runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCacheTest.java

```diff
@@ -0,0 +1,130 @@
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.client.grpc.stubs;
+
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Function;
+import org.apache.beam.runners.dataflow.worker.windmill.WindmillServiceAddress;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ChannelCacheTest {
+
+  private ChannelCache cache;
+
+  private static ChannelCache newCache(
+      Function<WindmillServiceAddress, ManagedChannel> channelFactory) {
+    return ChannelCache.forTesting(channelFactory, () -> {});
+  }
+
+  @After
+  public void cleanUp() {
+    if (cache != null) {
+      cache.clear();
+    }
+  }
+
+  private ManagedChannel newChannel(String channelName) {
+    return WindmillChannelFactory.inProcessChannel(channelName);
+  }
+
+  @Test
+  public void testLoadingCacheReturnsExistingChannel() {
+    String channelName = "existingChannel";
+    ManagedChannel channel = newChannel(channelName);
+    Function<WindmillServiceAddress, ManagedChannel> channelFactory =
+        spy(
+            new Function<WindmillServiceAddress, ManagedChannel>() {
+              @Override
+              public ManagedChannel apply(WindmillServiceAddress windmillServiceAddress) {
+                return channel;
+              }
+            });
+
+    cache = newCache(channelFactory);
+    WindmillServiceAddress someAddress = mock(WindmillServiceAddress.class);
+    // Initial call to load the cache.
+    cache.get(someAddress);
+
+    ManagedChannel cachedChannel = cache.get(someAddress);
+    assertSame(channel, cachedChannel);
+    verify(channelFactory, times(1)).apply(eq(someAddress));
+  }
+
+  @Test
+  public void testLoadingCacheReturnsLoadsChannelWhenNotPresent() {
+    String channelName = "existingChannel";
+    ManagedChannel channel = newChannel(channelName);
+    Function<WindmillServiceAddress, ManagedChannel> channelFactory =
+        spy(
+            new Function<WindmillServiceAddress, ManagedChannel>() {
+              @Override
+              public ManagedChannel apply(WindmillServiceAddress windmillServiceAddress) {
+                return channel;
+              }
+            });
+
+    cache = newCache(channelFactory);
+    WindmillServiceAddress someAddress = mock(WindmillServiceAddress.class);
+    ManagedChannel cachedChannel = cache.get(someAddress);
+    assertSame(channel, cachedChannel);
+    verify(channelFactory, times(1)).apply(eq(someAddress));
+  }
+
+  @Test
+  public void testRemoveAndClose() throws InterruptedException {
```

Comment:
see other comments
- test that a new call for the address gets a different channel not closed
- test that removal fn is async and doesn't block remove()

Reply:
done